drivers/pwm/pwm_numaker.c

```diff
@@ -0,0 +1,579 @@
+		/* data->capture[channel].is_busy = true; */
+	}
+}
+
+static void pwm_numaker_isr(const struct device *dev, uint32_t st_channel, uint32_t end_channel)
+{
+	const struct pwm_numaker_config *cfg = dev->config;
+	struct pwm_numaker_data *data = dev->data;
+	EPWM_T *epwm = cfg->epwm;
+	struct pwm_numaker_capture_data *capture;
+	uint32_t int_status;
+	uint32_t cap_intsts;
+	int i;
+	uint32_t int_mask = (BIT(st_channel) | BIT(end_channel));
+	uint32_t cap_int_rise_mask, cap_int_fall_mask;
+	uint32_t cap_int_mask =
+		(EPWM_CAPIF_CFLIF0_Msk << st_channel | EPWM_CAPIF_CRLIF0_Msk << st_channel |
+		 EPWM_CAPIF_CFLIF0_Msk << end_channel | EPWM_CAPIF_CRLIF0_Msk << end_channel);
+
+	/* Get Output int status */
+	int_status = epwm->AINTSTS & int_mask;
+	/* Clear Output int status */
+	if (int_status != 0x00) {
+		epwm->AINTSTS = int_status;
+	}
+
+	/* Get CAP int status */
+	cap_intsts = epwm->CAPIF & cap_int_mask;
+
+	/* PWM counter is timing critical, to avoid print msg from irq_isr
+	 *  until getting capture cycles.
+	 */
+	LOG_DBG("Channel=0x%x, CAPIEN=0x%x, CAPIF=0x%x, capIntMask=0x%x",
+		st_channel, epwm->CAPIEN, epwm->CAPIF, cap_int_mask);
+	if (cap_intsts != 0x00) { /* Capture Interrupt */
+		/* Clear CAP int status */
+		epwm->CAPIF = cap_intsts;
+
+		/* Rising latch or Falling latch */
+		for (i = st_channel; i <= end_channel; i++) {
+			capture = &data->capture[i];
+			if (capture == NULL) {
+				continue;
+			}
+			cap_int_rise_mask = (EPWM_CAPTURE_INT_RISING_LATCH << i);
+			cap_int_fall_mask = (EPWM_CAPTURE_INT_FALLING_LATCH << i);
+			if ((cap_int_rise_mask | cap_int_fall_mask) & cap_intsts) {
+				pwm_numaker_channel_cap(dev, epwm, i);
+			}
+		}
+	}
+}
+
+static void pwm_numaker_p0_isr(const struct device *dev)
+{
+	/* Pair0 service channel 0, 1 */
+	pwm_numaker_isr(dev, 0, 1);
+}
+
+static void pwm_numaker_p1_isr(const struct device *dev)
+{
+	/* Pair1 service channel 2, 3 */
+	pwm_numaker_isr(dev, 2, 3);
+}
+
+static void pwm_numaker_p2_isr(const struct device *dev)
+{
+	/* Pair2 service channel 4, 5 */
+	pwm_numaker_isr(dev, 4, 5);
+}
+#endif /* CONFIG_PWM_CAPTURE */
+
+/* PWM driver registration */
+static const struct pwm_driver_api pwm_numaker_driver_api = {
+	.set_cycles = pwm_numaker_set_cycles,
+	.get_cycles_per_sec = pwm_numaker_get_cycles_per_sec,
+#ifdef CONFIG_PWM_CAPTURE
+	.configure_capture = pwm_numaker_configure_capture,
+	.enable_capture = pwm_numaker_enable_capture,
+	.disable_capture = pwm_numaker_disable_capture,
+#endif /* CONFIG_PWM_CAPTURE */
+};
+
+/* Alternative EPWM clock get rate before support standard clock_control_get_rate */
+static int __pwm_numaker_clk_get_rate(EPWM_T *epwm, uint32_t *rate)
+{
+	uint32_t clk_src;
+	uint32_t epwm_clk_src;
+
+	if (epwm == EPWM0) {
+		clk_src = CLK->CLKSEL2 & CLK_CLKSEL2_EPWM0SEL_Msk;
+	} else if (epwm == EPWM1) {
+		clk_src = CLK->CLKSEL2 & CLK_CLKSEL2_EPWM1SEL_Msk;
+	} else {
+		LOG_ERR("Invalid EPWM node");
+		return -EINVAL;
+	}
```

Comment:
Non blocking: If you later decide to create `EPWM_GetFreq()`, you could also eliminate this pointer check by adding the clock source bit as a devicetree property.

Reply:
Yes, it will be a good choice for object validation.